### PR TITLE
Introducing Generic Truth Value

### PR DESCRIPTION
**Pull request by leungmanhin**

### CODE REVIEW

File: opencog/atomspace/GenericTruthValue.cc [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <math.h>
+#include "GenericTruthValue.h"
+
+using namespace opencog;
+
+#define KKK 800.0f
+#define CVAL  0.2f
```
Comment (linas) [inline opencog/atomspace/GenericTruthValue.cc]: Plese remove both of these, they should not be used.

File: opencog/atomspace/GenericTruthValue.cc [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+GenericTruthValue::GenericTruthValue(count_t pe, count_t te,
+                                     strength_t f, strength_t fs,
+                                     confidence_t c, entropy_t e)
+{
+    positiveEvidence = pe;
+    totalEvidence = te;
```
Comment (linas) [inline opencog/atomspace/GenericTruthValue.cc]: are these counts? or averages? I would like to see them called "count" if they are meant to integer-like .. e.g. positive EvidenceCount or just positiveCount.

File: opencog/atomspace/GenericTruthValue.cc [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+count_t GenericTruthValue::getTotalEvidence() const
+{
+    return totalEvidence;
+}
+
+count_t GenericTruthValue::getLogTotalEvidence() const
+{
+    return log(totalEvidence);
```
Comment (linas) [inline opencog/atomspace/GenericTruthValue.cc]: This should be log base 2, not the natural log.   Its enough to divide by 0.6931471805599453 to get this.

File: opencog/atomspace/GenericTruthValue.cc [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+confidence_t GenericTruthValue::getConfidence() const
+{
+    return confidence;
+}
+
+confidence_t GenericTruthValue::getLogConfidence() const
+{
+    return log(confidence);
```
Comment (linas) [inline opencog/atomspace/GenericTruthValue.cc]: Not sure that log confidence, or that log fuzzy strength even make sense ...

File: opencog/atomspace/GenericTruthValue.cc [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+GenericTruthValuePtr GenericTruthValue::merge(GenericTruthValuePtr gtv) const
+{
+    auto other_te = gtv->getTotalEvidence();
+    auto new_pe = positiveEvidence + gtv->getPositiveEvidence();
+    auto new_te = totalEvidence + other_te
+                  - std::min(totalEvidence, other_te) * CVAL;
+    auto new_f = (frequency * totalEvidence + gtv->getFrequency() * other_te)
+                 / (totalEvidence + other_te);
+    auto new_fs = std::max(fuzzyStrength, gtv->getFuzzyStrength());
+    auto new_c = new_te / (new_te + KKK);
```
Comment (linas) [inline opencog/atomspace/GenericTruthValue.cc]: Ughh. These formulas are seriously ugly. The formulas for frequency and total evidence are clearly wrong if one is just counting.   Confidence would also be wrong in a neural net activation model.  My gut instinct is that we should not be enshrining special formulas here; the user should provide the formulas needed for merging ...

File: opencog/atomspace/GenericTruthValue.cc (hunk 3 above)
Comment (ngeiswei) [inline opencog/atomspace/GenericTruthValue.cc]: you can also use log2 in /<COGUTILS_REPO/>/opencog/util/numeric.h

**Pull request merged**